NixOS friendly build process

jkozak opened the pull request:
1. keep given PYTHONPATH around (after our pip install dir)
2. build binaries rather than downloading

I'm not particularly expert in NixOS and there may be a better way of doing this, but at least these changes are minimal.  There is a performance hit for 2 but not much as a fraction of the overall build.

Tested on debian testing and NixOS 21.05

Commit:
NixOS friendly build process

1. keep given PYTHONPATH around (after our pip install dir)
2. build binaries rather than downloading

nazar-pc commented on worker/Makefile:
```diff
@@ -42,7 +42,7 @@ ifeq ($(wildcard $(PIP_DIR)),)
 		echo "Installation failed, likely because PIP is unavailable, if you are on Debian/Ubuntu or derivative please install the python3-pip package"
 	# Install `meson` and `ninja` using `pip` into custom location, so we don't depend on system-wide installation.
-	$(PYTHON) -m pip install --upgrade --target=$(PIP_DIR) meson ninja
+	$(PYTHON) -m pip install --upgrade --target=$(PIP_DIR) --no-binary ":all:" meson ninja
```
Why is this necessary though, can you provide an error output?
Also I believe this deserves a comment on why are we not using binaries.

BTW, you might be able to run command as is and fall back to compilation only when it fails.

jkozak replied:
NixOS doesn't have the standard `/usr` directories but has per-package bundles in a content-addressable store.  NixOS tweaks build processes to set the binary's `RPATH` to deal with this: and so binaries compiled elsewhere almost certainly won't work.

Could try the fallback thing - would need to do `ldd <binary>` and check for `not found` on a list of binaries.

Could also check explicitly for NixOS (existence of `/etc/NIXOS`); would need to do this for `guix` as well (and maybe others?).

nazar-pc replied:
Just do `x || y || z` and add a comment why we have the alternative option.

jkozak replied:
That will successfully create a ninja binary that doesn't work (on NixOS, GUIX)

nazar-pc replied:
Hm, you're right. Are NixOS folks aware of broken Python dependencies because of this at all? Maybe you can provide a link to the relevant issue?

jkozak replied:
https://nixos.wiki/wiki/Packaging/Binaries

It's not a bug it's a feature ;-)

nazar-pc replied:
I might have not been very clear. I suggested to report Python issue, maintainers of those distributions better make `--no-binary ":all:"` the default behavior of packages Python version.

jkozak replied:
Sorry, I misunderstood.  I will pass this very sensible suggestion on to the NixOS maintainers.

nazar-pc replied:
Will you be able to update this PR to have a check for NixOS and only build stuff on it (with comment to upstream issue so we can remove it in the future when NixOS fixes this problem)?

jkozak replied:
Will do.  Is `make` guaranteed to be GNU Make?  (it uses GNUisms at the moment, I think)

nazar-pc replied:
Yes, we target GNU Make right now

Commit:
NixOS friendly build process - explicitly detect NixOS and guix

This patch may be removed if

  https://github.com/NixOS/nixpkgs/issues/142383

and a similar change to guix is actioned.

nazar-pc reviewed (approved):
Looks good to me, but I would still like to see link to upstream report about this

Commit:
NixOS friendy build process: see associated bug #678

nazar-pc committed:
Add note with link to description why workaround is needed

nazar-pc commented on worker/Makefile:
```diff
@@ -14,6 +14,9 @@ DOCKER ?= docker
 MESON_ARGS ?= ""
+# Workaround for NixOS and Guix that don't work with pre-built binaries, see:
+# https://github.com/NixOS/nixpkgs/issues/142383.
+PIP_BUILD_BINARIES = $(shell [ -f /etc/NIXOS -o -d /etc/guix ] && echo "--no-binary :all:")
```
I just noticed this doesn't account for Windows at all. We almost certainly don't want to build things on Windows.

UPD: My bad, it does, everything is good.

nazar-pc commented:
Thank you, this will be included into the next release
